cityofthebigshoulders.game.php

```diff
@@ -4707,14 +4706,59 @@ function sellShares($selected_shares)
 
-                        self::notifyAllPlayers( "shareSold", "", array(
-                            'type' => $preferred_stock['card_type'],
-                            'id' => $preferred_stock_id,
-                            'player_id' => $new_owner_id
-                        ) );
+                            self::notifyAllPlayers( "shareTransferred", "", array(
+                                'type' => $preferred_stock['card_type'],
+                                'id' => $preferred_stock_id,
+                                'player_id' => $player_id,
+                                'from_id' => $new_owner_id
+                            ) );
+
+                            $number_to_give_back = 2;
+                            foreach($company['stocks_by_player'][$player_id]['stocks'] as $old_stock)
+                            {
+                                $can_give_back = true;
+                                foreach($stocks as $stock)
+                                {
+                                    if ($stock && $old_stock['card_id'] == $stock['card_id'])
+                                    {
+                                        $can_give_back = false; // already solded
+                                        break;
+                                    }
+                                }
+
+                                if ($can_give_back)
+                                {
+                                    $old_stock_id = $old_stock['card_id'];
+
+                                    self::DbQuery("UPDATE card SET card_location='bank', owner_type='bank', card_location_arg=0 WHERE card_id=$old_stock_id");
+
+                                    self::notifyAllPlayers( "shareSold", "", array(
+                                        'type' => $old_stock['card_type'],
+                                        'id' => $old_stock_id,
+                                        'player_id' => $player_id
+                                    ) );
+
+                                    $number_to_give_back--;
+                                    if ($number_to_give_back == 0)
+                                        break;
+                                }
+                            }
+                        }
+                        else
+                        {
+                            self::DbQuery("UPDATE card SET card_location='bank', owner_type='bank', card_location_arg=0 WHERE card_id=$preferred_stock_id");
 
+                            self::notifyAllPlayers( "shareSold", "", array(
+                                'type' => $preferred_stock['card_type'],
+                                'id' => $preferred_stock_id,
+                                'player_id' => $new_owner_id
+                            ) );
+                        }
                         $stock_count--;

```

Comment:
There is only one thing missing:

You need to update the stock count for the player selling the director's share in this case. I assumed that the number of certificates wouldn't change for the player selling, but in this case I was wrong.

If the player has 60% (4 certificates) and sells the director's share, then he will now have 30% with the preferred certificate (3 certificates).

In this special case, you can just add a counter for "current_shares_${player_id}" that is one less (it will always be the case).

Good work!

Reply:
I think it's done later in code
![image](https://user-images.githubusercontent.com/70916547/93721174-12407a80-fb5c-11ea-9795-e68e5e6545a1.png)

Reply:
I don't even remember the code I wrote 😄